experiments/templatetags/experiments.py

```diff
@@ -85,26 +111,263 @@ def experiment(parser, token):
+        lineno = parser.stream.current.lineno
+        args = [nodes.ContextReference()]
+        goal_name = parser.stream.current
+        args.append(self._name_or_const(goal_name))
+        next(parser.stream)
+        node = self.call_method(
+            'render_experiment_goal', args, lineno=lineno)
+        return nodes.CallBlock(node, [], [], []).set_lineno(lineno)
+
+    def render_experiment_goal(self, context, goal_name, caller):
+        """Callback to render {% experiment_goal ... %} tags"""
+        tmplt = template.loader.get_template('experiments/goal.html')
+        context_dict = dict(context)
+        context_dict.update(
+            _experiment_goal(goal_name)
+        )
+        return tmplt.render(context_dict)
+
+    def parse_experiment_enroll(self, parser):
+        """Parse {% experiment_enroll ... %} tags"""
+
+        lineno = parser.stream.current.lineno
+
+        # list of nodes that will be used when calling the callback:
+        args = []
+
+        # parsing first parameter:
+        experiment_name = parser.stream.current
+        args.append(self._name_or_const(experiment_name))
+        next(parser.stream)
+
+        # parsing remaining parameters (the "alternatives"):
+        alternatives = []
+        while parser.stream.current.type != 'block_end':
+            if self._token_as(parser):
+                break
+            alternatives.append(self._name_or_const(parser.stream.current))
+            next(parser.stream)
+        args.append(nodes.List(alternatives))
+
+        # expecting `as` after the alternatives:
+        if not self._token_as(parser):
+            raise TemplateSyntaxError(
+                'Syntax should be like: '
+                '{% experiment_enroll "experiment_name"'
+                ' "alternative1" "alternative2" ... as some_variable %}',
+                lineno,
+            )
+        next(parser.stream)
+
+        # parse what comes after `as`:
+        target = parser.parse_assign_target()
+
+        # We're done with parsing the tag.
+
+        # we will also need the context in the callback:
+        args.append(nodes.ContextReference())
+
+        # create a callback node that will be executed on render:
+        call_node = self.call_method(
+            'render_experiment_enroll', args, lineno=lineno)
+
+        # return an assignment node that will trigger the callback:
+        return nodes.Assign(target, call_node, lineno=lineno)
+
+    def render_experiment_enroll(self, experiment_name, alternatives, context):
+        """
+        Callback to render {% experiment_enroll ... %} tags.
+
+        This method does not actually render anything, but is called at
+        render time so keeping the name for consistency.
+        Result gets added ("assigned") back into template context.
+        """
+        return _experiment_enroll(context, experiment_name, *alternatives)
+
+    # helpers #
+
+    def _name_or_const(self, token):
```

Comment:
is it difficult to  seperate this function? I would prefer to have a couple of functions seperated for each case.

Reply:
I don't understand, can explain what should be separated?

Reply:
We have a function that returns `nodes.Name(token.value, 'load')` or `nodes.Const(token.value)`, depending on `token.type`. Maybe I am not clear about when is this function used.

Reply:
That's right, the function returns one or the other, depending on which token is passed in. Used in cases like this:

1) Say that `goal_name` is passed from Python view, i.e. it is in template's context. Then in the template we can use that variable like so:
`{% experiment_goal goal_name %}`

2) If the goal name can be hard-coded in the template (which is probably the more common usage), then we can simply write:
`{% experiment_goal "some_goal" %}` (notice the quotes)

In first example, `parser.stream` will contain a `Token` of type `name`, and in the second example it will be type `string`.

Same is true for `{% experiment_enroll %}` tag, except that for this tag any argument before `as` can be a variable (i.e. a "name"). (The one argument after `as` _must_ be a "name", but that's unrelated).

So instead of `if`-ing around multiple times while parsing those tags, I'm calling this helper method.